serverless/serverless
# Use zip command to package functions

## What did you implement:

Fixes #3054

In order to reduce memory consumption of packaging large projects, serverless will try to zip functions using available `zip` command. Otherwise fallback to `archiver` npm module as it was before. 

Benchmarks: https://github.com/serverless/serverless/issues/3054#issuecomment-272641038

## How did you implement it:

In two words: spawn `zip` binary to create an artifact. Fallback to old approach with `archiver` otherwise.

Strange moment: I couldn't get `zip` command understand glob patterns used in `sls`. So I decided to keep max backwards compatibility with `archiver` and use `glob-all` for both. More on this in the issue: https://github.com/serverless/serverless/issues/3054#issuecomment-272666235

The idea is to pass a list of files one by one to `zip`. But there comes an issue - we can easily have 10s of thousands of files. So it will exceed max string length allowed for any kind of command in unix.

This was overcame by splitting file paths into chunks, meaning  `zip` command could be  called several times for one archive, but subsequent calls just append data, so it works perfectly even though it sounds for me very hacky. 

I adjusted expectations about number of files in tests. This isn't a hack to "make tests green" but because previously there was `fs.createWriteStream(artifactFilePath)` before invoking `glob` file search. And it returned freshly created empty `.serverless/artifact.zip` as in tests we don't bother excluding `.serverless` directory. 

What more about tests, it's a bit hard to write unit tests for this file. I used to stubbing/mocking `require`s a lot, but I'm not sure if it's a good idea to create new files in the repo whenever I want. Please shed some light on the project structure rules or ideas how this can be tested better. Maybe zipping logic can be extracted into a new npm module.

I'm open to any kind of suggestions!

## How can we verify it:

* Please try it on different platforms as I have access to Linux only
* Basically `serverless deploy --noDeploy`
* In order to notice larger difference, try packaging on this example repo - https://github.com/pmuens/serverless-service-zip-testing + do the `npm install *some large libs*` for a more real-life example 
* Make sure it creates same file structure, packages exotic filenames and so on.

Also mentioning  @ryansb as he volunteered in testing

## Todos:



***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
